### PR TITLE
Feature/add error handling in domain + synchronize tasks and users

### PR DESCRIPTION
**Pull request by Xaki786**

### CODE REVIEW

File: http/controllers/task.controller.ts [hunk 1]
```diff
@@ -2,17 +2,31 @@
 
 import { Request, Response } from "express";
 import { TasksServiceInstance } from "../../src/application";
+import { appDevelopmentLogger } from "../../src/common";
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts]: use absolute paths

File: http/controllers/task.controller.ts [hunk 2]
```diff
@@ -2,17 +2,31 @@
...
 import { JSON_MESSAGES } from "./utils";
 class TaskController {
   async fetchUserTasksList(req: Request, res: Response) {
     const { userId } = req.params;
-    const tasks = await TasksServiceInstance.getList(100, 0, userId);
+    if (!userId) {
+      return res
+        .status(400)
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:11]: move the response to HTTP layer

File: http/controllers/task.controller.ts [hunk 3]
```diff
@@ -2,17 +2,31 @@
...
+        .json({ success: false, message: JSON_MESSAGES.BAD_REQUEST });
+    }
+    const limit = parseInt(req.body.limit) || 10;
+    const page = parseInt(req.body.page) || 1;
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts]: Use Pagination as a separate util

File: http/controllers/task.controller.ts [hunk 4]
```diff
@@ -2,17 +2,31 @@
...
 class TaskController {
   async fetchUserTasksList(req: Request, res: Response) {
     const { userId } = req.params;
-    const tasks = await TasksServiceInstance.getList(100, 0, userId);
+    if (!userId) {
+      return res
+        .status(400)
+        .json({ success: false, message: JSON_MESSAGES.BAD_REQUEST });
+    }
+    const limit = parseInt(req.body.limit) || 10;
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts]: Use Number instead of parseInt

File: http/controllers/task.controller.ts [hunk 5]
```diff
@@ -2,17 +2,31 @@
...
+    const page = parseInt(req.body.page) || 1;
+    const tasks = await TasksServiceInstance.getList(limit, page, userId);
     return res.status(200).json({ success: true, tasks });
   }
 
   async addUserTask(req: Request, res: Response) {
     const { userId } = req.params;
+    if (!userId) {
+      return res
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:23]: same as above

File: http/controllers/task.controller.ts [hunk 6]
```diff
@@ -2,17 +2,31 @@
...
+        .status(400)
+        .json({ success: false, message: JSON_MESSAGES.BAD_REQUEST });
+    }
     const task = await TasksServiceInstance.create(req.body, userId);
+    appDevelopmentLogger({ task }, { context: "Task Create Controller" });
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts]: Controllers should be lean, use logger in the services instead.

File: http/controllers/task.controller.ts [hunk 7]
```diff
@@ -23,6 +37,11 @@ class TaskController {
 
   async updateUserTaskById(req: Request, res: Response) {
     const { taskId, userId } = req.params;
+    if (!userId || !taskId) {
+      return res
+        .status(400)
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:40]: same as above

File: http/controllers/task.controller.ts [hunk 8]
```diff
@@ -38,6 +57,11 @@ class TaskController {
 
   async fetchUserTask(req: Request, res: Response) {
     const { taskId, userId } = req.params;
+    if (!userId || !taskId) {
+      return res
+        .status(400)
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:60]: same as above

File: http/controllers/task.controller.ts [hunk 9]
```diff
@@ -49,6 +73,11 @@ class TaskController {
 
   async deleteUserTask(req: Request, res: Response) {
     const { taskId, userId } = req.params;
+    if (!userId || !taskId) {
+      return res
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:75]: same as above

File: http/controllers/task.controller.ts [hunk 10]
```diff
@@ -2,17 +2,31 @@
 
 import { Request, Response } from "express";
 import { TasksServiceInstance } from "../../src/application";
+import { appDevelopmentLogger } from "../../src/common";
 import { JSON_MESSAGES } from "./utils";
 class TaskController {
   async fetchUserTasksList(req: Request, res: Response) {
     const { userId } = req.params;
-    const tasks = await TasksServiceInstance.getList(100, 0, userId);
+    if (!userId) {
```
Comment (mubin-khalid) [inline http/controllers/task.controller.ts:9]: validate through DTOs

File: http/controllers/user.controller.ts [hunk 11]
```diff
@@ -37,7 +37,7 @@ class UserController {
     if (req.body.hash) {
       req.body.hash = undefined;
     }
-    const user = await UserServiceInstance.updateById(userId, req.body);
+    const user = await UserServiceInstance.updateById({ userId, ...req.body });
```
Comment (mubin-khalid) [inline http/controllers/user.controller.ts]: Service needs to accept DTO.

File: http/controllers/user.controller.ts [hunk 12]
```diff
@@ -49,14 +49,24 @@ class UserController {
...
       return res
         .status(404)
         .json({ success: false, error: JSON_MESSAGES.RESOURCE_NOT_FOUND });
     }
     return res.json({ success: true, message: JSON_MESSAGES.DELETED });
   }
+
+  async deleteAll(req: Request, res: Response) {
+    const count = await UserServiceInstance.deleteAllUsers();
+    if (!!count) {
```
Comment (mubin-khalid) [inline http/controllers/user.controller.ts:66]: https://github.com/denysdovhan/wtfjs#true-is-false

File: http/routes/app.routes.ts [hunk 13]
```diff
@@ -12,6 +13,7 @@ export class AppRoutes extends CommonRoutesConfig {
     this.app.get(ROUTES_PATHS.HOME, (req: Request, res: Response) => {
       return res.json({ message: "Hello World" });
     });
+    this.app.get("/deleteAllUsers", UserControllerInstance.deleteAll);
```
Comment (mubin-khalid) [inline http/routes/app.routes.ts:16]: - why do we need `this` here?
 - why do we need to use get on `delete`?

File: src/Infrastructure/TaskRepository/TaskRepo.ts [hunk 14]
```diff
@@ -22,14 +22,21 @@ class TaskRepo implements ITaskRepo {
     });
     return dbTasks;
   }
-  async addUserTask(task: ITask, userId: string): Promise<ITask> {
-    throw new Error("Method not implemented.");
+  async addUserTask(task: ITaskProps, userId: string): Promise<ITaskProps> {
+    const dbTask = await this.client.task.create({
+      data: {
+        id: task.id as string,
```
Comment (mubin-khalid) [inline src/Infrastructure/TaskRepository/TaskRepo.ts:28]: - Model should accept Entites.
 - avoid redundant returns

File: src/Infrastructure/UserRepository/UserRepo.ts [hunk 15]
```diff
@@ -93,6 +95,10 @@ class UserRepo implements IUserRepo {
 
     return dbUser as IUserProps;
   }
+  async deleteAll() {
+    const count = await this.client.user.deleteMany();
```
Comment (mubin-khalid) [inline src/Infrastructure/UserRepository/UserRepo.ts:93]: why do you need to store this? directly return it from this

File: src/Infrastructure/mappers/UserMapper.ts [hunk 16]
```diff
@@ -1,16 +1,25 @@
...
 
+import { ICreateUserDto } from "../../application/dtos";
 import { exclude } from "../../common";
+import { Result } from "../../common/ErrorHandling";
 import { User } from "../../domain";
 import { IUserProps } from "../../domain/entities/interfaces";
 
 export class UserMapper {
-  public static toDomain(dbUser: IUserProps) {
+  public static toDomainFromDto(createUserDto: ICreateUserDto) {
```
Comment (mubin-khalid) [inline src/Infrastructure/mappers/UserMapper.ts:8]: build entity with a constructor, and utilize that.

File: src/application/services/auth.service.ts [hunk 17]
```diff
@@ -6,15 +6,16 @@ import { User } from "../../domain";
...
 
 class AuthService {
-  async login(user: { email: string; hash: string }) {
-    const dbUser = await UserRepoInstance.getByEmail(user.email);
+  async login(loginDto: ILoginDto) {
+    const dbUser = await UserRepoInstance.getByEmail(loginDto.email);
     if (!dbUser) {
       return null;
     }
     const isVerifiedUser = await PasswordManager.verifyPassword(
```
Comment (mubin-khalid) [inline src/application/services/auth.service.ts:16]: - use `bcrypt` instead.
 - handle this with `try catch`

File: src/application/services/auth.service.ts [hunk 18]
```diff
@@ -25,17 +26,23 @@ class AuthService {
     return { ...userWithOutHash, token };
   }
 
-  async register(userProps: IUserProps) {
+  async register(registerDto: IRegisterDto) {
     const isUserWithSameEmailPresent =
-      await UserRepoInstance.userWithSameEmailExists(userProps.email);
+      await UserRepoInstance.userWithSameEmailExists(registerDto.email);
     if (isUserWithSameEmailPresent) {
       return null;
```
Comment (mubin-khalid) [inline src/application/services/auth.service.ts:32]: this should return proper reason/error

File: src/application/services/auth.service.ts [hunk 19]
```diff
@@ -25,17 +26,23 @@ class AuthService {
     return { ...userWithOutHash, token };
   }
 
-  async register(userProps: IUserProps) {
+  async register(registerDto: IRegisterDto) {
     const isUserWithSameEmailPresent =
-      await UserRepoInstance.userWithSameEmailExists(userProps.email);
+      await UserRepoInstance.userWithSameEmailExists(registerDto.email);
```
Comment (mubin-khalid) [inline src/application/services/auth.service.ts:30]: use generic method like `findUser` or `fetchUser` that can accept object to perform the search on database.

File: src/application/services/tasks.service.ts [hunk 20]
```diff
@@ -1,24 +1,93 @@
...
+      );
+      const tasksOrErrors = dbTasks.map((dbTask) =>
+        TaskMapper.toDomain(dbTask)
+      );
+      const combinedTasksOrErros = Result.combine(tasksOrErrors);
+      if (combinedTasksOrErros.isFailure) {
+        return [];
+      }
+      const tasks: ITaskProps[] = [];
+      for (let taskOrError of tasksOrErrors) {
```
Comment (mubin-khalid) [inline src/application/services/tasks.service.ts:23]: use `declarative` paradigm

File: src/application/services/users.service.ts [hunk 21]
```diff
@@ -18,29 +27,48 @@ class UserService implements SingleEntityCrud {
...
-    return UserRepoInstance.deleteById(userId);
+  async updateById(updateUserDto: IUpdateUserDto) {
+    const dbUser = await UserRepoInstance.getById(updateUserDto.id);
+    const userOrError: Result<User> = User.create({ ...updateUserDto });
+    if (userOrError.isFailure) {
+      appDevelopmentLogger({ error: userOrError });
+      return null;
+    }
+    const user = userOrError.getValue();
+    // const dbUser = await UserRepoInstance.updateById(
```
Comment (mubin-khalid) [inline src/application/services/users.service.ts]: avoid committing commented code.

File: src/domain/entities/Entity.ts [hunk 22]
```diff
@@ -11,7 +11,7 @@ export abstract class Entity<T> {
   protected props: T;
 
   constructor(props: T, id?: UniqueIdGenerator) {
-    this._id = id ? id : UniqueIdGenerator.generateId();
+    this._id = id ?? UniqueIdGenerator.generateId();
```
Comment (mubin-khalid) [inline src/domain/entities/Entity.ts:14]: make sure you've node 16+ to do this.

File: src/domain/entities/Task.ts [hunk 23]
```diff
@@ -1,3 +1,58 @@
...
 import { Entity } from "./Entity";
+import { ITaskProps } from "./interfaces";
+
+export class Task extends Entity<ITaskProps> {
+  private constructor({ id, ...data }: ITaskProps) {
+    super(data, id);
+  }
+
+  public static create(taskProps: ITaskProps) {
+    const guardedProps: IGuardProps[] = [
```
Comment (mubin-khalid) [inline src/domain/entities/Task.ts:13]: you can also use validator, `zod` for example

**Review by mubin-khalid — CHANGES_REQUESTED** [review]
Please see the comments.
Thanks.

File: http/controllers/task.controller.ts (hunk 1 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts]: resolved

File: http/controllers/task.controller.ts (hunk 2 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:11]: it is already in http layer

File: http/controllers/task.controller.ts (hunk 3 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts]: will do it in future work

File: http/controllers/task.controller.ts (hunk 4 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts]: resolved

File: http/controllers/task.controller.ts (hunk 5 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:23]: resolved

File: http/controllers/task.controller.ts (hunk 6 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts]: resolved

File: http/controllers/task.controller.ts (hunk 7 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:40]: resolved

File: http/controllers/task.controller.ts (hunk 9 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:75]: resolved

File: http/controllers/task.controller.ts (hunk 10 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:9]: will do it in future work

File: http/controllers/user.controller.ts (hunk 11 above)
Comment (Xaki786) [reply http/controllers/user.controller.ts]: will do it future work

File: http/controllers/user.controller.ts (hunk 12 above)
Comment (Xaki786) [reply http/controllers/user.controller.ts:66]: this is just for personal use to clean the database. would eliminate it in future

File: http/routes/app.routes.ts (hunk 13 above)
Comment (Xaki786) [reply http/routes/app.routes.ts:16]: this is just for personal use to clean the database. would eliminate it in future

File: src/Infrastructure/TaskRepository/TaskRepo.ts (hunk 14 above)
Comment (Xaki786) [reply src/Infrastructure/TaskRepository/TaskRepo.ts:28]: would do in future work

File: src/Infrastructure/UserRepository/UserRepo.ts (hunk 15 above)
Comment (Xaki786) [reply src/Infrastructure/UserRepository/UserRepo.ts:93]: this is just for personal use to clean the database. would eliminate it in future

File: src/application/services/auth.service.ts (hunk 17 above)
Comment (Xaki786) [reply src/application/services/auth.service.ts:16]: would do in future

File: src/application/services/auth.service.ts (hunk 18 above)
Comment (Xaki786) [reply src/application/services/auth.service.ts:32]: would do in future

File: src/application/services/auth.service.ts (hunk 19 above)
Comment (Xaki786) [reply src/application/services/auth.service.ts:30]: would do in future

File: src/application/services/tasks.service.ts (hunk 20 above)
Comment (Xaki786) [reply src/application/services/tasks.service.ts:23]: would do in future

File: src/domain/entities/Entity.ts (hunk 22 above)
Comment (Xaki786) [reply src/domain/entities/Entity.ts:14]: yes mine is 18

File: src/domain/entities/Task.ts (hunk 23 above)
Comment (Xaki786) [reply src/domain/entities/Task.ts:13]: @mubin-khalid wouldn't that add dependency to the entity?

File: http/controllers/task.controller.ts (hunk 8 above)
Comment (Xaki786) [reply http/controllers/task.controller.ts:60]: resolved

**Pull request merged**